[Cases] Use the new internal users API in the UI

cnasikas opened the pull request:
## Summary

This PR updates the UI to use the new `users` API introduced in https://github.com/elastic/kibana/pull/149663. I changed the backend response to accommodate the needs of the UI.

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

cnasikas commented on x-pack/plugins/cases/common/types.ts:
```diff
@@ -9,10 +9,16 @@ type SnakeToCamelCaseString<S extends string> = S extends `${infer T}_${infer U}
       [K in keyof T as SnakeToCamelCaseString<K & string>]: SnakeToCamelCase<T[K]>;
     }
+  : T extends unknown[]
+  ? SnakeToCamelCaseArray<T>
```
extends the utility function to support arrays.

cnasikas commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.tsx:
```diff
@@ -100,14 +128,6 @@ export const CaseViewActivity = ({
     [onUpdateField]
   );
 
-  const emailContent = useMemo(
```
Moved it inside the `UserList` component.

cnasikas commented on x-pack/plugins/cases/public/containers/use_find_case_user_actions.test.tsx:
```diff
@@ -80,137 +74,4 @@ describe('UseFindCaseUserActions', () => {
     expect(addError).toHaveBeenCalled();
   });
-
-  describe('getProfileUids', () => {
```
functionality got removed from the `useFindCaseUserActions` hook.

cnasikas commented on x-pack/plugins/cases/public/containers/use_find_case_user_actions.tsx:
```diff
@@ -5,64 +5,22 @@
-
-      return {
-        caseUserActions,
-        participants,
-        profileUids,
```
The information is returned by the new users API

cnasikas commented on x-pack/plugins/cases/server/client/user_actions/users.ts:
```diff
@@ -51,14 +51,17 @@ export const getUsers = async (
     ]);
 
-    const userProfiles = await getUserProfiles(securityStartPlugin, userProfileUids);
+    const userProfiles = await getUserProfiles(securityStartPlugin, userProfileUids, 'avatar');
```
The UI needs the `avatar` data to show the avatars correctly.

cnasikas commented on x-pack/plugins/cases/server/client/user_actions/users.ts:
```diff
@@ -51,14 +51,17 @@ export const getUsers = async (
       .map((participant) => participant.user.profile_uid) as string[];
 
     const assigneesUids = theCase.case.assignees.map((assignee) => assignee.uid);
+    const reporter = theCase.case.created_by;
```
The reporter will be part of the users API response.

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
Pinging @elastic/response-ops-cases (Feature:Cases)

jonathan-buttner commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.tsx:
```diff
@@ -209,20 +229,22 @@ export const CaseViewActivity = ({
-            userProfiles={userProfiles}
-          />
-          {userActionsData?.participants ? (
+          {caseUsers?.reporter ? (
```
Should we show the `Unknown` user if we don't have the reporter instead of hiding the list?

I suppose another option would be to fallback to the case data right?

cnasikas replied:
Thanks! I will fallback to the case data.

jonathan-buttner commented on x-pack/plugins/cases/public/containers/mock.ts:
```diff
@@ -957,3 +958,85 @@ export const getPersistableStateAttachment = (
 });
+
+export const getCaseUsersMockResponse = (): CaseUsers => {
+  return {
```
Should we have a mock response that only has the `profileUid`?

jonathan-buttner reviewed (commented):
Code changes look good, it looks like the UI is only showing the emails now though

<img width="1025" alt="image" src="https://user-images.githubusercontent.com/56361221/217654711-9b51f1ae-0706-474b-b3a8-b5665696f72e.png">

Question: Does this PR switch the participants to include the currently assigned users? In the currently assigned users aren't show in the participants list.

js-jankisalvi commented:
> Code changes look good, it looks like the UI is only showing the emails now though

I also see the same behaviour. Showing `username` instead of `email` will look cleaner

cnasikas commented:
> Code changes look good, it looks like the UI is only showing the emails now though
> 
> 
> Question: Does this PR switch the participants to include the currently assigned users? In the currently assigned users aren't show in the participants list.

> I also see the same behaviour. Showing username instead of email will look cleaner

Damn, Rigid Rodent did it again. Thanks, both of you for finding the bug. I should be more careful. I pushed the fix and added many unit tests to check user profile variations. The bug occurred because I converted the response to camel case when the UI expected it to be in snake case. That was the format returned by the `bulkGet` of the security plugin. I slightly modified the backend response to be more inline with the response from the security plugin to avoid in the future this kind of bugs.

cnasikas commented on x-pack/plugins/cases/public/containers/mock.ts:
```diff
@@ -957,3 +958,85 @@ export const getPersistableStateAttachment = (
+        },
+      },
+    ],
+    unassignedUsers: [
```
There are no stupid questions 🙂! Participants are all users that did an action in the case. For example, edit a description, a comment, etc. Now, if you assign a user to a case a user action will be created containing the uid of the user. Same if you remove an assignee (unassigned user). We need the profile information for that users.

js-jankisalvi replied:
ahh, got it. Thanks 🙂

cnasikas commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.tsx:
```diff
@@ -33,6 +33,40 @@ import { useFindCaseUserActions } from '../../../containers/use_find_case_user_a
+    ...users.assignees,
+    ...users.participants,
+    users.reporter,
+    ...users.unassignedUsers,
```
I think the `Map` below guarantees that only unique profiles will be added. The map will contain all valid user profiles.

js-jankisalvi commented:
> Damn, Rigid Rodent did it again. Thanks, both of you for finding the bug. I should be more careful. I pushed the fix and added many unit tests to check user profile variations. The bug occurred because I converted the response to camel case when the UI expected it to be in snake case. That was the format returned by the `bulkGet` of the security plugin. I slightly modified the backend response to be more inline with the response from the security plugin to avoid in the future these kind of bugs. Could you please also test that the avatar is displayed correctly if you change it to be an image or a different color? You can do that by going to:


Updated user profile with image, Avatar icon is displayed:
![Screenshot 2023-02-09 at 14 25 58](https://user-images.githubusercontent.com/117571355/217826015-eebe5f25-61f9-4121-b814-4c9995fc6481.png)


Tested again. Looks good now. (Note: Reporter shows old profile image of elastic user)
![image](https://user-images.githubusercontent.com/117571355/217826763-2d74025c-2a5d-4f94-8acc-b74b140d456d.png)

js-jankisalvi reviewed (approved):
Verified locally, looks great 👍 🎉

jonathan-buttner commented on x-pack/plugins/cases/common/api/user.ts:
```diff
@@ -7,19 +7,22 @@
-    user: UserRt,
+    user: UserWithoutProfileUidRt,
   }),
+  rt.partial({ uid: rt.string }),
```
I wonder if there's a way we can tie this implementation with the security plugin 🤔 My concern is that if they make a breaking change to the response structure and make the changes throughout the code base, this won't show up as needing to be changed and it will break the UI.

Here's an idea, how about we add an integration test that does a bulk get using the security API and then does a decode using this schema? Or something like that, that way if there's a breaking change at least we'll get a failing test.

cnasikas replied:
Amazing idea! I was thinking about the same but I could not find a good way to do it.

jonathan-buttner commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.tsx:
```diff
@@ -229,22 +235,20 @@ export const CaseViewActivity = ({
               headline={i18n.PARTICIPANTS}
               loading={isLoadingUserActions}
-              users={caseUsers.participants}
+              users={[...caseUsers.participants, ...caseUsers.assignees]}
```
Does the ternary operator on line 245 need to be updated to reflect the `assignees` too?

How about we combine the array above this component and just check to see if it's not empty? Although is it even possible to have assignees but not a participant 🤔 that'd mean we don't have the reporter's information I suppose. Either probably best to check just in case the backend has a bug.

jonathan-buttner commented on x-pack/test/cases_api_integration/security_and_spaces/tests/common/internal/user_actions_get_users.ts:
```diff
@@ -167,204 +169,256 @@ export default ({ getService }: FtrProviderContext): void => {
+                full_name: userProfile.user.full_name ?? null,
+                username: userProfile.user.username ?? null,
+              },
+              avatar,
```
I think we want the `userProfile.data.avatar` here right to ensure that it's returned in the format we expect?

jonathan-buttner reviewed (approved):
Great work!
